### PR TITLE
Fix DeleteOnExitHook cause memory leak

### PR DESCRIPTION
**Pull request by wuxiansen**
Motivation:

If DeleteOnExitHook is in the open state and the program runs for a long time, the DeleteOnExitHook file keeps increasing.
This results in a cause memory leak

 

See https://github.com/netty/netty/issues/10351

Modification:

I re-customized a DeleteOnExitHook hook. If DeleteOnExitHook is turned on, this hook will be added when creating a temporary file. After the request ends and the corresponding resources are released, the current file will be removed from this hook, so that it will not increase all the time.

Result:

Fixes #[10351](https://github.com/netty/netty/issues/10351).

### CODE REVIEW

**Comment by fredericBregier** [comment]
LGTM

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.io.File;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Extend {@link java.io.DeleteOnExitHook} to dynamically manipulate hook records.
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:26]: I'm not able to make sense of this comment. We don't actually extend that class.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public final class DeleteFileOnExitHook {
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:28]: Does this class really need to be public?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+     * @param file tmp file path
+     */
+    public static synchronized void remove(String file) {
+        try {
+            if (files != null) {
+                files.remove(file);
+            }
+        } catch (Exception e) {
+            logger.warn("The cleanup file path failed.", e);
+        }
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:61]: I don't think there's any exception that can happen here?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        synchronized (DeleteFileOnExitHook.class) {
+            toBeDeleted = files;
+            files = null;
+        }
+
+        if (!toBeDeleted.isEmpty()) {
+            for (String filename : toBeDeleted) {
+                (new File(filename)).delete();
+            }
+        }
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:89]: With a Set based on ConcurrentHashMap, we can just iterate it directly. Since we are at JVM shut down at this point, I don't think we need to clear the set.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        final String dir = "target/DeleteFileOnExitHookTest/customBaseDirAndDeleteOnHookExit";
+        File baseDir = new File(dir);
+        baseDir.mkdirs();  // we don't need to clean it since it is in volatile files anyway
+
+        defaultHttpDataFactory.setBaseDir(dir);
+        defaultHttpDataFactory.setDeleteOnExit(true);
+        final Attribute attr = defaultHttpDataFactory.createAttribute(req1, "attribute1");
+        final FileUpload fu = defaultHttpDataFactory.createFileUpload(
+                req1, "attribute1", "tmp_f.txt", "text/plain", null, null, 0);
+        //DeleteFileOnExitHook.add(new File(dir, fu.getFilename()).getAbsolutePath());
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:49]: Remove comment.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        assertEquals(dir, DiskAttribute.class.cast(attr).getBaseDirectory());
+        assertEquals(dir, DiskFileUpload.class.cast(fu).getBaseDirectory());
+        assertTrue(DiskAttribute.class.cast(attr).deleteOnExit());
+        assertTrue(DiskFileUpload.class.cast(fu).deleteOnExit());
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:54]: Easier to do the casts at variable declaration site.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test DeleteFileOnExitHook
+ */
+public class DeleteFileOnExitHookTest {
+    private static final HttpRequest req1 = new DefaultHttpRequest(HTTP_1_1, POST, "/form");
+
+    @Test
+    public void customBaseDirAndDeleteOnHookExit() throws IOException {
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:38]: What behaviour are we trying to verify?

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        //DeleteFileOnExitHook.add(new File(dir, fu.getFilename()).getAbsolutePath());
+
+        assertEquals(dir, DiskAttribute.class.cast(attr).getBaseDirectory());
+        assertEquals(dir, DiskFileUpload.class.cast(fu).getBaseDirectory());
+        assertTrue(DiskAttribute.class.cast(attr).deleteOnExit());
+        assertTrue(DiskFileUpload.class.cast(fu).deleteOnExit());
+
+        fu.setContent(Unpooled.wrappedBuffer(new byte[]{1, 2, 3, 4}));
+        assertTrue(fu.getFile().exists());
+        //fu.delete();
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:58]: Remove comment.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        assertEquals(0, fu.getFile().length());
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:60]: Why is the length zero when we set 4 bytes of content?

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        fu.delete();
+        //fu.release();
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:62]: Remove comment. Also, we should verify that the file no longer exists, and that it's pathname is no longer in the set of files to delete on shut down.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    }
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:63]: We should also have a test that verifies that the registered files are deleted by the shut down hook.

**Review by chrisvest — CHANGES_REQUESTED** [review]
Thanks for submitting this PR! I had some comments, please take a look.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.io.File;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Extend {@link java.io.DeleteOnExitHook} to dynamically manipulate hook records.
+ */
+public final class DeleteFileOnExitHook {
+    private static Set<String> files = new HashSet<String>();
```
[the remark this answers is not in the document]
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:29]: Good idea.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 1 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:26]: You right.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 2 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:28]: At first I thought maybe it would be used elsewhere.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 3 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:61]: Maybe it is redundant，If an unknown problem occurs with `add` and `remove`, I just want to make sure that it will not affect the normal process. 

Or am I thinking a little too much?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 4 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:89]: Ok

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java (hunk 9 above)
Comment (wuxiansen) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:60]: Throw an exception to test whether the hook will execute properly at the end of the program.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java (hunk 7 above)
Comment (wuxiansen) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:38]: Okay, I got it.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 3 above)
Comment (chrisvest) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:61]: Yeah, they're just methods on Set. They can only throw Errors or RuntimeExceptions, and I think we should just let those bubble out if they ever happen.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java (hunk 11 above)
Comment (wuxiansen) [reply codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:63]: Ok

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:2]: nit: 2020

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 2 above)
Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:28]: please declare it package-private

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 3 above)
Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:61]: I agree with @chrisvest here..

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2017 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:2]: 2020

File: transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java [hunk 15]
```diff
@@ -93,7 +93,10 @@ private void handleReadException(ChannelPipeline pipeline, ByteBuf byteBuf, Thro
         allocHandle.readComplete();
         pipeline.fireChannelReadComplete();
         pipeline.fireExceptionCaught(cause);
-        if (close || cause instanceof IOException) {
+
+        // If oom will close the read event, release connection.
+        // See https://github.com/netty/netty/issues/10434
+        if (close || cause instanceof OutOfMemoryError || cause instanceof IOException) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java:99]: This change doesn't seem to be related to this pr ... please revert

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java [hunk 16]
```diff
@@ -123,7 +123,10 @@ private void handleReadException(ChannelPipeline pipeline, ByteBuf byteBuf, Thro
             allocHandle.readComplete();
             pipeline.fireChannelReadComplete();
             pipeline.fireExceptionCaught(cause);
-            if (close || cause instanceof IOException) {
+
+            // If oom will close the read event, release connection.
+            // See https://github.com/netty/netty/issues/10434
+            if (close || cause instanceof OutOfMemoryError || cause instanceof IOException) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:129]: This change doesn't seem to be related to this pr ... please revert

File: transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java [hunk 17]
```diff
@@ -587,7 +587,10 @@ private void handleReadException(ChannelPipeline pipeline, ByteBuf byteBuf, Thro
                 allocHandle.readComplete();
                 pipeline.fireChannelReadComplete();
                 pipeline.fireExceptionCaught(cause);
-                if (close || cause instanceof IOException) {
+
+                // If oom will close the read event, release connection.
+                // See https://github.com/netty/netty/issues/10434
+                if (close || cause instanceof OutOfMemoryError || cause instanceof IOException) {
```
Comment (normanmaurer) [inline transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java:593]: This change doesn't seem to be related to this pr ... please revert

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java [hunk 18]
```diff
@@ -726,7 +726,10 @@ private void handleReadException(ChannelPipeline pipeline, ByteBuf byteBuf, Thro
             allocHandle.readComplete();
             pipeline.fireChannelReadComplete();
             pipeline.fireExceptionCaught(cause);
-            if (close || cause instanceof IOException) {
+
+            // If oom will close the read event, release connection.
+            // See https://github.com/netty/netty/issues/10434
+            if (close || cause instanceof OutOfMemoryError || cause instanceof IOException) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java:732]: This change doesn't seem to be related to this pr ... please revert

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 3 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:61]: Okay, I've adjusted it.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 2 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:28]: OK.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java (hunk 18 above)
Comment (wuxiansen) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java]: Ok

File: transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java (hunk 17 above)
Comment (wuxiansen) [reply transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java]: Ok

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java (hunk 16 above)
Comment (wuxiansen) [reply transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: Ok

File: transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java (hunk 15 above)
Comment (wuxiansen) [reply transport/src/main/java/io/netty/channel/oio/AbstractOioByteChannel.java:99]: Ok

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 13 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java]: ok

**Comment by wuxiansen** [comment]
@chrisvest @normanmaurer  I've adjusted it. 😊

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+     * check in records.
+     *
+     * @param file target file
+     * @return true or false
+     */
+    public static boolean checkFileExist(String file) {
+        return FILES.contains(file);
+    }
+
+
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:74]: nit remove empty line

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by wuxiansen** [comment]
Done.
Review is a good thing.  👍 @normanmaurer

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Clean up all the files.
+     */
+    private static void runHooks() {
+        for (String filename : FILES) {
+            new File(filename).delete();
+        }
+    }
+}
```
Comment (chrisvest) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:82]: Nit: make sure files end with a line-break.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+        fu.setContent(Unpooled.wrappedBuffer(new byte[]{1, 2, 3, 4}));
+        assertTrue(fu.getFile().exists());
+
+        String filePath = fu.getFile().getPath();
+        assertTrue(DeleteFileOnExitHook.checkFileExist(filePath));
+
+        fu.release();
+        assertFalse(DeleteFileOnExitHook.checkFileExist(filePath));
+    }
+}
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:85]: Nit: another line break, please.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+    @Test
+    public void testDeleteFileOnExitHookExecutionSuccessful() {
+        File[] files = new File(HOOK_TEST_TMP).listFiles(new FilenameFilter() {
+            @Override
+            public boolean accept(File dir, String name) {
+                return name.startsWith(DiskFileUpload.prefix);
+            }
+        });
+
+        assertEquals(0, files.length);
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:62]: Tests don't run in any particular order, so this test also needs to do the work to ensure that some file uploads are created. It can then call `runHooks` directly (since it has package visibility) to simulate the shutdown.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+    @Test
+    public void testAfterHttpDataReleaseCheckFileExist() throws IOException {
+        final DefaultHttpDataFactory defaultHttpDataFactory = new DefaultHttpDataFactory(true);
+        File baseDir = new File(HOOK_TEST_TMP);
+        baseDir.mkdirs();  // we don't need to clean it since it is in volatile files anyway
+
+        defaultHttpDataFactory.setBaseDir(HOOK_TEST_TMP);
+        defaultHttpDataFactory.setDeleteOnExit(true);
+        final FileUpload fu = defaultHttpDataFactory.createFileUpload(
+                REQUEST, "attribute1", "tmp_f.txt", "text/plain", null, null, 0);
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:74]: This set-up work could be extracted to a test helper method.

File: codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Test
+    public void testTriggerDeleteFileOnExitHook() throws IOException {
+        final DefaultHttpDataFactory defaultHttpDataFactory = new DefaultHttpDataFactory(true);
+        File baseDir = new File(HOOK_TEST_TMP);
+        baseDir.mkdirs();  // we don't need to clean it since it is in volatile files anyway
+
+        defaultHttpDataFactory.setBaseDir(HOOK_TEST_TMP);
+        defaultHttpDataFactory.setDeleteOnExit(true);
+        final FileUpload fu = defaultHttpDataFactory.createFileUpload(
```
Comment (chrisvest) [inline codec-http/src/test/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHookTest.java:46]: This variable doesn't need to be final.

**Review by chrisvest — COMMENTED** [review]
Almost there. Just a few small things.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 20 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java]: Done.

**Review by normanmaurer — APPROVED** [review]

**Review by chrisvest — APPROVED** [review]

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 25]
```diff
@@ -71,10 +71,12 @@ public static boolean checkFileExist(String file) {
...
 
-
-    static void runHooks() {
+    /**
+     * Clean up all the files.
+     */
+    private static void runHooks() {
         for (String filename : FILES) {
-            (new File(filename)).delete();
+            new File(filename).delete();
```
Comment (johnou) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:79]: Try catch ignore exception?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+    static {
+        // DeleteOnExitHook must be the last shutdown hook to be invoked.
+        // Application shutdown hooks may add the first file to the
+        // delete on exit list and cause the DeleteOnExitHook to be
+        // registered during shutdown in progress.
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+
+            @Override
+            public void run() {
+                runHooks();
```
Comment (johnou) [inline codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:41]: run deletion hook?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 25 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:79]: If the file does not exist, return false .

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 26 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:41]: 😵，  Change `runHooks` method name?

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 25 above)
Comment (johnou) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:79]: security manager still might throw an exception or NPE if null was passed as path.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 26 above)
Comment (johnou) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:41]: right, eg. runDeletionHook, plural hooks doesn't make sense because it is only one hook.

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 25 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:79]: `ConcurrentHashMap` key not be null.  so avoided this problem

File: codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java (hunk 26 above)
Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:41]: Good advice ! 😊

Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java:41]: did you forgot to push ?

Comment (wuxiansen) [reply codec-http/src/main/java/io/netty/handler/codec/http/multipart/DeleteFileOnExitHook.java]: > did you forgot to push ?

⊙﹏⊙‖∣     

 Done.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Review by johnou — APPROVED** [review]

**Comment by normanmaurer** [comment]
@wuxiansen did you sign our ICLA ? https://netty.io/s/icla ?

**Comment by wuxiansen** [comment]
> @wuxiansen did you sign our ICLA ? https://netty.io/s/icla ?

Yes, I have applied for it not long ago.

**Comment by normanmaurer** [comment]
@wuxiansen thanks a lot!

**Comment by wuxiansen** [comment]
> @wuxiansen thanks a lot!

my pleasure.🤝
